### ISSUE TITLE
Recoil::execute when/how can it be used?

### ISSUE THREAD

**Issue by djsharman** [recoilphp/recoil#111]
Hi James,

I am still experimenting with my Database layer and trying to write async DataObject code in a synchronous way.

I wrote some test code show below, but it doesn't do what I would expect. I am starting to wonder if I really understand generators at all (functions that can generate values multiple times AFAIK?). Anyway I wrote the code below as I was seeing Recoil::execute functions not being called at all when called from Ratchet messages. I am trying to simulate the same thing here using a timer. Interestingly Timer 2 never fires at all, can you explain why that is? Is there something obvious I am missing here?

Cheers,

Darren

```php
use Recoil\Kernel\Strand;
use Recoil\React\ReactKernel;
use Recoil\Recoil;
use React\EventLoop\Factory;


$loop = \React\EventLoop\Factory::create();




$kernel = new ReactKernel($loop);
$kernel->execute( function () use ($loop, $kernel) {
	
	$i = 0;
	$loop->addPeriodicTimer(0.1, function() use (&$i) {
		echo '1 Timer'.++$i, PHP_EOL;
	});
	
	$j = 0;
	$timer = $loop->addPeriodicTimer(0.1, function() use (&$j, $kernel) {
		print '2Timer'.++$j. PHP_EOL;
		
		$strand = yield Recoil::execute(function () {
			print 'Async 1' . PHP_EOL;
			yield;
		
		});
		
		$kernel->waitForStrand($strand);
			
		$strand2 = yield Recoil::execute(function ()  {
	
			print 'Async 2' . PHP_EOL;
			yield;
		});
	
		$kernel->waitForStrand($strand2);
		
	});
	
    
    yield 0;
	
});


$kernel->wait();

```

**Comment by jmalloc** [comment]
Hi Darren,

[This post](https://medium.com/async-php/co-operative-php-multitasking-ce4ef52858a0#.2c76suf6b) provides an *excellent* introduction to generators, and how they can be used as coroutines. 

The problem with your example is that ReactPHP's timers don't know how to execute generators as coroutines, so when the timer handlers fire, the generator they produce is simply discarded.

There are a couple of different ways to do what you're doing in Recoil:

### Example 1 - using loops instead of timers

This example doesn't (directly) use the React event loop at all. I suppose you could consider this "idiomatic" Recoil, since this example would work the same on any future Kernel implementation (e.g. #85), not just the ReactPHP one.

```php
$kernel = new ReactKernel($loop);
$kernel->execute(function () {
    // Start a new strand that runs in a loop, equivalent to a periodic timer.
    yield Recoil::execute(function () {
        $i = 0;
        while (true) {
            yield 0.1;
            echo 'Timer #1, Tick ', ++$i, PHP_EOL;
        }
    });

    yield Recoil::execute(function() {
        $i = 0;
        while (true) {
            yield 0.1;
            echo 'Timer #2, Tick ', ++$i, PHP_EOL;

            $strand = yield Recoil::execute(function () {
                echo 'Async 1' . PHP_EOL;
                yield;
            });

            // Yielding a strand blocks this coroutine until the yielded strand exits.
            yield $strand;

            $strand = yield Recoil::execute(function () {
                echo 'Async 2' . PHP_EOL;
                yield;
            });

            yield $strand;
        }
    });
});


$kernel->wait();
```

### Example 2 - using `Recoil::callback()` with ReactPHP timers

This example uses `Recoil::callback()`, which takes a coroutine function as a parameter and produces a "normal" function that starts that coroutine in a new strand. `Recoil::callback()` is intended to help integrate with non-Recoil asynchronous code, so this isn't the way I'd go about writing new code, but nonetheless I've included it here for completeness:

```php
$kernel = new ReactKernel($loop);
$kernel->execute(function () {
    // Get the loop using the Recoil Kernel API, this method is only available
    // under the ReactKernel implementation.
    $loop = yield Recoil::eventLoop();

    $i = 0;
    $loop->addPeriodicTimer(0.1, yield Recoil::callback(function () use (&$i) {
        echo 'Timer #1, Tick ', ++$i, PHP_EOL;
        yield Recoil::execute(/*... etc*/);
    }));
});

$kernel->wait();
```

Please be aware I just wrote the examples up in the browser, so they may not execute as is, but they should be pretty close, I hope this is helpful to you :)

**Comment by djsharman** [comment]
Hi James, 

Thank you again for your help. I had seen that article on introduction to generators, that is how I found your project . :-)

Please take a look at the code below. It works, but not how I would like it too. 

On the line  `$x = yield $Obj->go();  //<<<<<<<<<<<<<< here`

I need to use yield to get the coroutine to execute. However what I want to do is get the value of the generator and not have to yield. How is this possible?

I have tried:

$x = $Obj->go();  = coroutine doesn't run.
$x = $Obj->go()->current();  = I get an API object back.

I hope what I am trying to do is clear. I want the coroutine to run async, then get the resulting value before my code continues. (Trying to be async with logically written synchronous looking code).

A couple of other questions. 
- Should I close the issues I raised, or do you do that?
- Are you going to make a forum, or community around this project? I can't be the only person who is asking these kind of questions, and the answers will get lost if left in the GitHub Issues list. 

Cheers,

Darren

`
use Recoil\Kernel\Strand;
use Recoil\React\ReactKernel;
use Recoil\Recoil;
use React\EventLoop\Factory;




$kernel = new ReactKernel();
$kernel->execute( function () use ($kernel) {
	
	yield Recoil::execute(function () {
		$i = 0;
		while (true) {
			yield 0.1;
			echo 'Timer #1, Tick ', ++$i, PHP_EOL;
		}
	});
	
		yield Recoil::execute(function() {

			$Obj = new Tester();
			$x = yield $Obj->go();  //<<<<<<<<<<<<<< here
			$x++;			
			print "Value is: $x" . PHP_EOL;
			
			$strand = yield Recoil::execute(function () {
				echo 'Async 2' . PHP_EOL;
				yield;
			});
			
			yield $strand;

		});
	
	
});


	class Tester {
		public function go() {
			print "In function GO!\n";
			
			$x = 1;
	
			$strand =  yield Recoil::execute(function () use (&$x) {
				echo 'Async 1' . PHP_EOL;
				
				$x = 3.14159;
                yield;
					
			});
			echo 'Starting Wait 1' . PHP_EOL;
			yield $strand;
			echo 'Finished Wait 1' . PHP_EOL;


			print "Leaving function GO!\n";

			
			return $x;

		}
	}


$kernel->wait();

`

**Comment by jmalloc** [comment]
Please feel free close any issues that you feel I've addressed. In regards to a forum, I hadn't really thought about it. Questions tend to get asked via email, but I definitely prefer people to use GitHub issues as then the info is publicly available. Mostly however, Recoil is used internally at my day job so I haven't yet had the need for anything more public. I will definitely be addressing the lack of documentation.

---

When you say ...

> It works, but not how I would like it too.

... do you mean that you'd expect the output from the looping "timer" strand to be interleaved with the output from the others?

If that's not the case let me know, but I'll move forward with that assumption for now :)

> I need to use yield to get the coroutine to execute. However what I want to do is get the value of the generator and **not have to yield**. How is this possible?

The short answer is that you can't do this without yielding - but I'm not sure why that's a concern. Allow me to waffle on for a bit, and forgive me if I'm telling you what you already know ...

`yield` is the language feature that Recoil "exploits" in order to get control of the current coroutine. Wherever you see `yield` in a coroutine, you are giving control to the kernel to do something. This means that yielding is perfectly normal and does not necessarily imply any synchronous or asynchronous behaviour.

The expression `yield $func()` means "call `$func` as a coroutine on the current strand", so you're using it appropriately for the behaviour you desire. `$func` is executed until it completes, and the `yield` expression evaluates to its return value (or, if `$func` throws an exception, it propagates "out of" the `yield`). This is Recoil's equivalent of calling a normal PHP function.

Contrast this behaviour with that of `yield Recoil::execute($func)`, which runs `$func` on a new strand, and **continues executing the current coroutine**. It is guaranteed that the current coroutine continues before the new one is started. There is no equivalent of this in vanilla PHP, but it's kind of like using React's `futureTick()`.

With all of that said, it's important to remember that PHP is (generally) not multithreaded. React and/or Recoil do not give you true concurrency. The upshot of this is that *you* need to tell Recoil when to "cooperate", that is, when to allow CPU time for other strands to execute. This will happen naturally if your application is IO bound, but there are several ways you can do this explicitly:

1. Perform IO with `yield Recoil::read()`, `write()` or `select()`
1. Sleep with `yield Recoil::sleep($seconds)` (or `yield $seconds`)
1. Explicitly "cooperate" with `yield Recoil::cooperate()` (or just `yield;`) - you might use this inside a loop of CPU-bound code.

It may not be obvious, but React and similar libraries work exactly this way, but instead of `sleep` you'd use a timer, and instead of `cooperate` you'd use `futureTick`.

> I hope what I am trying to do is clear. I want the coroutine to run async, then get the resulting value before my code continues. (Trying to be async with logically written synchronous looking code).

Other than the sleep operation in the loop, there is no asynchronous operation being performed in your example. Let's emulate one using another `sleep` ...

```php
// This is our "asynchronous" operation. It pretends to run an SQL query that
// takes some time by sleeping, but performing IO (as you would when sending a
// real query to the server), would work the same ...
function runQuery($sql) {
    print "Entered runQuery()!" . PHP_EOL;
    yield 2; // OMG this query is slow it takes two seconds.
    print "Leaving runQuery()!" . PHP_EOL;

    return 3.141593;
}

$kernel->execute(function () {
    // Start the "timer" strand ...
    yield Recoil::execute(function () {
        while (true) {
            yield 0.25;
            echo '(tick)' . PHP_EOL;
        }
    });

    // Run our query. Note that the "timer" strand continues to tick while we
    // wait for the result ...
    echo 'Running query ...' . PHP_EOL;
    $pi = yield runQuery('SELECT PI()');
    echo 'PI is ' . $pi . PHP_EOL;
});

$kernel->wait();
```

In summary:

- `yield` does not imply "cooperating". You must sleep, perform IO, or explicitly cooperate.
- Strands started with `Recoil::execute()` execute *after* the calling coroutine is resumed, meaning that the calling strand must "cooperate" in one of the ways mentioned above before the new strand is given time to execute. This is why the output from the timer strand in your example appeared after that of the `go()` method.
- You should rarely need to use the kernel directly, or explicitly wait for a strand to complete.

Again, I hope this helps. I've tried to provide as much info as I could because our differing timezones make it hard to have a straightforward discussion. If any of the terminology is confusing or I simply haven't addressed the issue, please let me know :)
